addons/io/org.openhab.io.transport.modbus/src/main/java/org/openhab/io/transport/modbus/internal/ModbusManagerImpl.java

```diff
@@ -0,0 +1,910 @@
+                    EndpointPoolConfiguration endpointPoolConfig = new EndpointPoolConfiguration();
+                    endpointPoolConfig.setInterTransactionDelayMillis(DEFAULT_TCP_INTER_TRANSACTION_DELAY_MILLIS);
+                    endpointPoolConfig.setConnectMaxTries(Modbus.DEFAULT_RETRIES);
+                    return endpointPoolConfig;
+                }
+
+                @Override
+                public @NonNull EndpointPoolConfiguration visit(ModbusSerialSlaveEndpoint modbusSerialSlavePoolingKey) {
+                    EndpointPoolConfiguration endpointPoolConfig = new EndpointPoolConfiguration();
+                    // never "disconnect" (close/open serial port) serial connection between borrows
+                    endpointPoolConfig.setReconnectAfterMillis(-1);
+                    endpointPoolConfig.setInterTransactionDelayMillis(DEFAULT_SERIAL_INTER_TRANSACTION_DELAY_MILLIS);
+                    endpointPoolConfig.setConnectMaxTries(Modbus.DEFAULT_RETRIES);
+                    return endpointPoolConfig;
+                }
+
+                @Override
+                public @NonNull EndpointPoolConfiguration visit(ModbusUDPSlaveEndpoint modbusUDPSlavePoolingKey) {
+                    EndpointPoolConfiguration endpointPoolConfig = new EndpointPoolConfiguration();
+                    endpointPoolConfig.setInterTransactionDelayMillis(DEFAULT_TCP_INTER_TRANSACTION_DELAY_MILLIS);
+                    endpointPoolConfig.setConnectMaxTries(Modbus.DEFAULT_RETRIES);
+                    return endpointPoolConfig;
+                }
+            });
+        });
+
+        GenericKeyedObjectPool<ModbusSlaveEndpoint, ModbusSlaveConnection> genericKeyedObjectPool = new ModbusConnectionPool(
+                connectionFactory);
+        genericKeyedObjectPool.setSwallowedExceptionListener(new SwallowedExceptionListener() {
+
+            @SuppressWarnings("null")
+            @Override
+            public void onSwallowException(@Nullable Exception e) {
+                LoggerFactory.getLogger(ModbusManagerImpl.class).error(
+                        "Connection pool swallowed unexpected exception: {}",
+                        Optional.ofNullable(e).map(ex -> ex.getMessage()).orElse("<null>"));
+            }
+        });
+        connectionPool = genericKeyedObjectPool;
+        this.connectionFactory = connectionFactory;
+    }
+
+    private Optional<ModbusSlaveConnection> borrowConnection(ModbusSlaveEndpoint endpoint) {
+        Optional<ModbusSlaveConnection> connection = Optional.empty();
+        KeyedObjectPool<ModbusSlaveEndpoint, ModbusSlaveConnection> pool = connectionPool;
+        if (pool == null) {
+            return connection;
+        }
+        long start = System.currentTimeMillis();
+        try {
+            connection = Optional.ofNullable(pool.borrowObject(endpoint));
+        } catch (Exception e) {
+            logger.warn("Error getting a new connection for endpoint {}. Error was: {} {}", endpoint,
```

Comment:
Shouldn't you rather throw an exception to the caller here instead of simply continuing?

Reply:
Again, up to you. In general I think it is better to let callers know if something goes wrong and let it react to it. Warn loggings from inner code makes the user to be the one who needs to react and this should be avoided when ever possible. Especially io-transport bundles are considered to be internal framework libraries, and how to deal with problems in connectivity etc. should imho always be left to the caller (i.e. the application logic which also interfaces with the user). So having any visible logging (error/warn/info) from this layer should usually only be done in case of unexpected situations, i.e. a software bug that should be reported.

Reply:
The current architecture in this transport bundle is then perhaps more "higher level" than others since the way connection errors are handled is dictated by the transport bundle. Currently there is configurable retries, with configurable sleep period between retries.

In fact, one the main features of this transport bundle is taking care all related to connections, "serializing requests" and ensuring only one connection is open at a time.

It would be a large change (I think) to "externalize" error handling since the requests are executed in background in a scheduled matter -- the normal java exception propagation is not enough.

I would like to keep this as is.

Reply:
> this transport bundle is then perhaps more "higher level" than others

This frightens me a bit as it might be a hint that there's too much logic in this layer already.
I'll refrain from looking deeper into it to avoid the risk of asking for bigger a bigger refactoring 🙈 .

Reply:
Oh no!

I tried to cover my plans already when I started the work 1.5 years ago (see [here](https://github.com/openhab/openhab2-addons/pull/362#issuecomment-293479264)) and actually thought [we were on the same page](https://github.com/openhab/openhab2-addons/pull/362#issuecomment-244693535).

But seriously speaking, I think this design should facilitiate the device specific bindings which I consider the main thing / most usual thing. We can always extend the API later to expose the details if necessary, right?

Furthermore, if the caller want total control -- why not just use the (3rd party modbus library) jamod directly? Currently I have not exposed any jamod stuff from the transport bundle to make it easier to change it afterwards to something that is actually maintained (for motivation, see [post regarding different libraries](https://github.com/openhab/openhab2-addons/pull/362#issuecomment-293617488)).

Reply:
Sorry, I didn't mean that as any kind of blame or criticism. It is just that we had meanwhile many discussions around the right level for the different libs for the MQTT integration and I think this is a pretty similar situation to modbus. And I think we ended up having a "lower-level" on the io-transport and more "connection broker" functionality in the generic MQTT binding (which defines the bridges that others can then use as connections).
But anyhow, I think there is no need to restructure anything here - you did a fantastic job 👍 .

Reply:
No offence taken really, sorry if I sounded harsh. I agree about the analogy to MQTT, it is very similar in nature to Modbus in some sense.

Thanks :)